src/charm.py

```diff
@@ -0,0 +1,451 @@
+        ).strip()
+
+        self.certificates.request_certificate_creation(certificate_signing_request=csr)
+
+    def _on_certificate_available(self, event: CertificateAvailableEvent) -> None:
+        """Handle the on certificate available event."""
+        if not self.peer_relation.app_data:
+            event.defer()
+            return
+        assert self.peer_relation.app_data.private_key
+        self.write_file(content=self.peer_relation.app_data.private_key, path=KEY_FILE_PATH)
+        self.write_file(content=event.certificate, path=CERT_FILE_PATH)
+        self.write_file(content=event.ca, path=CA_FILE_PATH)
+
+    @staticmethod
+    def write_file(content: str, path: str) -> None:
+        """Write content in a file."""
+        os.makedirs(os.path.dirname(path), exist_ok=True)
+        with open(path, "w") as f:
+            f.write(content)
+
+    def _start_process_action(self, event: ActionEvent) -> None:
+        """Handle the start process action."""
+        if self.peer_relation.unit_data.pid:
+            event.fail(f"Process id {self.peer_relation.unit_data.pid} already running!")
+
+        app_type = self.config.role
+        username = event.params["username"]
+        password = event.params["password"]
+        servers = event.params["servers"]
+        topic = event.params["topic_name"]
+        consumer_group_prefix = None
+        if "consumer_group_prefix" in event.params:
+            consumer_group_prefix = event.params["consumer_group_prefix"]
+
+        pid = self._start_process(
+            process_type=app_type,
+            username=username,
+            password=password,
+            servers=servers,
+            topic=topic,
+            consumer_group_prefix=consumer_group_prefix,
+        )
+        event.set_results({"process": pid})
+
+    def _stop_process_action(self, event: ActionEvent):
+        """Handle the stop process action."""
+        if not self.peer_relation.unit_data.pid:
+            event.fail("No process to stop!")
+            return
+
+        pid = self.peer_relation.unit_data.pid
+        r_pid = self._stop_process(pid)
+        assert r_pid == pid
+        event.set_results({"process": pid})
+
+    def _start_process(
+        self,
+        process_type: AppType,
+        username: str,
+        password: str,
+        servers: str,
+        topic: str,
+        consumer_group_prefix: Optional[str],
+    ) -> int:
+        """Handle the start of the process for consumption or production of messagges."""
+        t0 = int(time())
+        my_cmd = f"{self._build_cmd(process_type, username, password, servers, topic, consumer_group_prefix)}"
+        logger.info(my_cmd)
+        process = subprocess.Popen(
+            shlex.split(my_cmd),
+            stdout=open(f"/tmp/{t0}_{process_type.value}.log", "w"),
+            stderr=open(f"/tmp/{t0}_{process_type.value}.err", "w"),
+        )
+        self.peer_relation.set_pid(process.pid)
+        return process.pid
```

Comment:
I believe I tried something similar but I was actually stuck in this line for 10-20 minutes to make it work. I believe the two formulation are really equivalent from the complexity standpoint. But I would not change it, since, really, I'm afraid something could go wrong, and debugging this is of course a nightmare